tasks/indexing/task_indexing.go

```diff
@@ -0,0 +1,298 @@
+	RawSize   int64                   `db:"raw_size"`
+}
+
+func (i *IndexingTask) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+
+	var tasks []itask
+
+	ctx := context.Background()
+
+	err = i.db.Select(ctx, &tasks, `SELECT 
+											mit.id as id, 
+											mit.sp_id as sp_id, 
+											mit.sector_number as sector_number, 
+											mit.piece_cid as piece_cid, 
+											mit.piece_size as piece_size, 
+											mit.created_at as created_at, 
+											mit.piece_offset as piece_offset, 
+											mit.reg_seal_proof as reg_seal_proof,
+											md.chain_deal_id as chain_deal_id,
+											mk.file_size as raw_size
+										FROM 
+											market_indexing_tasks mit
+										JOIN 
+											market_mk12_deal_pipeline mk ON mit.id = mk.uuid
+										JOIN 
+											market_mk12_deals md ON mk.id = md.uuid
+										WHERE 
+											mit.task_id = $1 
+										ORDER BY 
+											mit.created_at ASC;`, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("getting indexing params: %w", err)
+	}
+
+	if len(tasks) != 1 {
+		return false, xerrors.Errorf("expected 1 sector params, got %d", len(tasks))
+	}
+
+	task := tasks[0]
+
+	// Check if piece is already indexed
+	var indexed bool
+	err = i.db.Select(ctx, &indexed, `SELECT indexed FROM market_piece_metadata WHERE piece_cid = $1`, task.PieceCid)
+	if err != nil {
+		return false, xerrors.Errorf("checking if piece is already indexed: %w", err)
+	}
+
+	// Return early if already indexed
+	if indexed {
+		err = i.recordCompletion(ctx, task, taskID)
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+
+	unsealed, err := i.pieceProvider.IsUnsealed(ctx, storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(task.SpID),
+			Number: task.Sector,
+		},
+		ProofType: task.Proof,
+	}, storiface.UnpaddedByteIndex(task.Offset), task.Size.Unpadded())
+	if err != nil {
+		return false, xerrors.Errorf("checking if sector is unsealed :%w", err)
+	}
+
+	if !unsealed {
+		return false, xerrors.Errorf("sector %d for miner %d is not unsealed", task.Sector, task.SpID)
+	}
+
+	pieceCid, err := cid.Parse(task.PieceCid)
+	if err != nil {
+		return false, xerrors.Errorf("parsing piece CID: %w", err)
+	}
+
+	reader, err := i.pieceProvider.ReadPiece(ctx, storiface.SectorRef{
+		ID: abi.SectorID{
+			Miner:  abi.ActorID(task.SpID),
+			Number: task.Sector,
+		},
+		ProofType: task.Proof,
+	}, storiface.UnpaddedByteIndex(task.Offset), abi.UnpaddedPieceSize(task.Size), pieceCid)
+
+	if err != nil {
+		return false, xerrors.Errorf("getting piece reader: %w", err)
+	}
+
+	recs := make([]indexstore.Record, 0)
+	opts := []carv2.Option{carv2.ZeroLengthSectionAsEOF(true)}
+	blockReader, err := carv2.NewBlockReader(reader, opts...)
+	if err != nil {
+		return false, fmt.Errorf("getting block reader over piece: %w", err)
+	}
+
+	blockMetadata, err := blockReader.SkipNext()
+	for err == nil {
+		recs = append(recs, indexstore.Record{
```

Comment:
so as to avoid bloating memory, can we add to the DB as we go, perhaps sending to a batcher thread?